File: scripts/merge-audio.js

import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const PROJECT_ROOT = path.resolve(__dirname, '..');
const INTRO_JINGLE = path.join(PROJECT_ROOT, 'public', 'music', 'intro_jingle.mp3');
const AUDIO_OUTPUT_DIR = path.join(PROJECT_ROOT, 'public', 'audio');
const SYNC_SCRIPT = path.join(PROJECT_ROOT, 'scripts', 'sync-audio.js');

function processFile(wavPath) {
    if (!fs.existsSync(wavPath)) {
        console.error(`Error: ${wavPath} not found.`);
        return;
    }

    // Clean up filename: remove .wav, _tts, and parenthetical trailing numbers like (1)
    let baseName = path.basename(wavPath, path.extname(wavPath));
    baseName = baseName.replace(/_tts/gi, '')
        .replace(/\s*\(\d+\)$/, '')
        .trim();

    const outputMp3 = path.join(AUDIO_OUTPUT_DIR, `${baseName}.mp3`);

    // Skip if already exists to avoid redundant processing and duplicates
    if (fs.existsSync(outputMp3)) {
        console.log(`Skipping: ${baseName}.mp3 already exists.`);
        return;
    }

    console.log(`Processing: ${wavPath} -> ${outputMp3}`);

    // FFmpeg command to merge intro + target + intro with 44.1kHz Stereo resampling
    try {
        const filterComplex =
            `[0:a]aresample=44100:osr=44100:ochl=stereo[a0];` +
            `[1:a]aresample=44100:osr=44100:ochl=stereo[a1];` +
            `[2:a]aresample=44100:osr=44100:ochl=stereo[a2];` +
            `[a0][a1][a2]concat=n=3:v=0:a=1[out]`;

        const cmd = `ffmpeg -y -i "${INTRO_JINGLE}" -i "${wavPath}" -i "${INTRO_JINGLE}" -filter_complex "${filterComplex}" -map "[out]" -b:a 192k "${outputMp3}"`;

        execSync(cmd, { stdio: 'inherit' });
        console.log(`Successfully created ${outputMp3}`);
    } catch (err) {
        console.error(`FFmpeg error processing ${wavPath}:`, err.message);
    }
}

function main() {
    const DOWNLOADS_DIR = path.join(process.env.USERPROFILE || 'C:\\Users\\admin', 'Downloads');

    const searchDirs = [PROJECT_ROOT, DOWNLOADS_DIR];
    let allWavFiles = [];

    console.log("Searching for WAV files in:");
    searchDirs.forEach(dir => {
        if (!fs.existsSync(dir)) return;
        console.log(` - ${dir}`);
        const files = fs.readdirSync(dir);
        const wavs = files
            .filter(f => f.toLowerCase().endsWith('.wav') && !f.toLowerCase().includes('jingle'))
            .map(f => path.join(dir, f));
        allWavFiles = allWavFiles.concat(wavs);
    });

    if (allWavFiles.length === 0) {
        console.log("No pending .wav files found in project root or Downloads.");
        return;
    }

    console.log(`Found ${allWavFiles.length} files to process.`);
    for (const wavPath of allWavFiles) {
        processFile(wavPath);
    }

    console.log("\nRunning sync-audio.js...");
    try {
        execSync(`node "${SYNC_SCRIPT}"`, { stdio: 'inherit' });
    } catch (err) {
        console.error("Error running sync-audio.js:", err.message);
    }

    console.log("\nAll files processed successfully.");
}

main();

File: scripts/merge-audio.py

import os
import subprocess
import sys
import glob

# Constants
PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
INTRO_JINGLE = os.path.join(PROJECT_ROOT, 'public', 'music', 'intro_jingle.mp3')
AUDIO_OUTPUT_DIR = os.path.join(PROJECT_ROOT, 'public', 'audio')
SYNC_SCRIPT = os.path.join(PROJECT_ROOT, 'scripts', 'sync-audio.js')

def process_file(wav_path):
    if not os.path.exists(wav_path):
        print(f"Error: {wav_path} not found.")
        return

    # Extract book id from wav file name (assuming format {id}_tts.wav or {id}.wav)
    base_name = os.path.basename(wav_path).replace('_tts.wav', '').replace('.wav', '')
    output_mp3 = os.path.join(AUDIO_OUTPUT_DIR, f"{base_name}.mp3")

    print(f"Processing: {wav_path} -> {output_mp3}")

    # FFmpeg command to merge intro + target + intro with 44.1kHz Stereo resampling
    # We use -filter_complex for concat and resampling
    cmd = [
        'ffmpeg', '-y',
        '-i', INTRO_JINGLE,
        '-i', wav_path,
        '-i', INTRO_JINGLE,
        '-filter_complex',
        '[0:a]aresample=44100:osr=44100:ochl=stereo[a0];'
        '[1:a]aresample=44100:osr=44100:ochl=stereo[a1];'
        '[2:a]aresample=44100:osr=44100:ochl=stereo[a2];'
        '[a0][a1][a2]concat=n=3:v=0:a=1[out]',
        '-map', '[out]',
        '-b:a', '192k',
        output_mp3
    ]

    try:
        subprocess.run(cmd, check=True)
        print(f"Successfully created {output_mp3}")
        
        # Run sync-audio.js
        print("Running sync-audio.js...")
        subprocess.run(['node', SYNC_SCRIPT], check=True)
    except subprocess.CalledProcessError as e:
        print(f"FFmpeg error: {e}")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    if len(sys.argv) > 1:
        process_file(sys.argv[1])
    else:
        # Look for all .wav files in root (excluding jingles)
        wav_files = [f for f in glob.glob(os.path.join(PROJECT_ROOT, "*.wav")) if "jingle" not in f.lower()]
        
        if not wav_files:
            print("No pendings .wav files found in project root.")
        else:
            print(f"Found {len(wav_files)} files to process.")
            for wav in wav_files:
                process_file(wav)
            print("\nAll files processed successfully.")
